### PR TITLE
optimizations

### PR DESCRIPTION
**Pull request by egonelbre**
Total improvement:

```
benchmark                                    old MB/s     new MB/s     speedup
BenchmarkMultiple/Encode/r30t60/60B-8        3.44         4.19         1.22x
BenchmarkMultiple/Decode/r30t60/60B-8        0.51         0.83         1.63x
BenchmarkMultiple/Encode/r30t60/120B-8       6.30         7.88         1.25x
BenchmarkMultiple/Decode/r30t60/120B-8       0.97         1.58         1.63x
BenchmarkMultiple/Encode/r30t60/240B-8       11.36        13.19        1.16x
BenchmarkMultiple/Decode/r30t60/240B-8       1.97         3.10         1.57x
BenchmarkMultiple/Encode/r30t60/510B-8       27.56        34.14        1.24x
BenchmarkMultiple/Decode/r30t60/510B-8       4.30         6.87         1.60x
BenchmarkMultiple/Encode/r30t60/1020B-8      48.27        59.25        1.23x
BenchmarkMultiple/Decode/r30t60/1020B-8      8.09         13.36        1.65x
BenchmarkMultiple/Encode/r30t60/255KB-8      296.98       316.09       1.06x
BenchmarkMultiple/Decode/r30t60/255KB-8      341.87       384.57       1.12x
BenchmarkMultiple/Encode/r30t60/1023KB-8     295.31       329.63       1.12x
BenchmarkMultiple/Decode/r30t60/1023KB-8     369.55       377.74       1.02x
BenchmarkMultiple/Encode/r30t60/5119KB-8     248.24       251.57       1.01x
BenchmarkMultiple/Decode/r30t60/5119KB-8     319.20       338.14       1.06x
BenchmarkMultiple/Encode/r30t60/8191KB-8     241.00       244.53       1.01x
BenchmarkMultiple/Decode/r30t60/8191KB-8     301.70       279.69       0.93x
```

Improvements from removing `map`:
```
benchmark                                    old MB/s     new MB/s     speedup
BenchmarkMultiple/Encode/r30t60/60B-8        3.44         3.88         1.13x
BenchmarkMultiple/Decode/r30t60/60B-8        0.51         0.75         1.47x
BenchmarkMultiple/Encode/r30t60/120B-8       6.30         7.01         1.11x
BenchmarkMultiple/Decode/r30t60/120B-8       0.97         1.51         1.56x
BenchmarkMultiple/Encode/r30t60/240B-8       11.36        11.85        1.04x
BenchmarkMultiple/Decode/r30t60/240B-8       1.97         2.77         1.41x
BenchmarkMultiple/Encode/r30t60/510B-8       27.56        30.16        1.09x
BenchmarkMultiple/Decode/r30t60/510B-8       4.30         6.24         1.45x
BenchmarkMultiple/Encode/r30t60/1020B-8      48.27        50.06        1.04x
BenchmarkMultiple/Decode/r30t60/1020B-8      8.09         11.99        1.48x
BenchmarkMultiple/Encode/r30t60/255KB-8      296.98       280.98       0.95x
BenchmarkMultiple/Decode/r30t60/255KB-8      341.87       356.18       1.04x
BenchmarkMultiple/Encode/r30t60/1023KB-8     295.31       292.75       0.99x
BenchmarkMultiple/Decode/r30t60/1023KB-8     369.55       365.20       0.99x
BenchmarkMultiple/Encode/r30t60/5119KB-8     248.24       236.62       0.95x
BenchmarkMultiple/Decode/r30t60/5119KB-8     319.20       333.59       1.05x
BenchmarkMultiple/Encode/r30t60/8191KB-8     241.00       225.91       0.94x
BenchmarkMultiple/Decode/r30t60/8191KB-8     301.70       305.00       1.01x
```

Improvements from SSE3 asm changes:
```
benchmark                                    old MB/s     new MB/s     speedup
BenchmarkMultiple/Encode/r30t60/60B-8        3.88         4.19         1.08x
BenchmarkMultiple/Decode/r30t60/60B-8        0.75         0.83         1.11x
BenchmarkMultiple/Encode/r30t60/120B-8       7.01         7.88         1.12x
BenchmarkMultiple/Decode/r30t60/120B-8       1.51         1.58         1.05x
BenchmarkMultiple/Encode/r30t60/240B-8       11.85        13.19        1.11x
BenchmarkMultiple/Decode/r30t60/240B-8       2.77         3.10         1.12x
BenchmarkMultiple/Encode/r30t60/510B-8       30.16        34.14        1.13x
BenchmarkMultiple/Decode/r30t60/510B-8       6.24         6.87         1.10x
BenchmarkMultiple/Encode/r30t60/1020B-8      50.06        59.25        1.18x
BenchmarkMultiple/Decode/r30t60/1020B-8      11.99        13.36        1.11x
BenchmarkMultiple/Encode/r30t60/255KB-8      280.98       316.09       1.12x
BenchmarkMultiple/Decode/r30t60/255KB-8      356.18       384.57       1.08x
BenchmarkMultiple/Encode/r30t60/1023KB-8     292.75       329.63       1.13x
BenchmarkMultiple/Decode/r30t60/1023KB-8     365.20       377.74       1.03x
BenchmarkMultiple/Encode/r30t60/5119KB-8     236.62       251.57       1.06x
BenchmarkMultiple/Decode/r30t60/5119KB-8     333.59       338.14       1.01x
BenchmarkMultiple/Encode/r30t60/8191KB-8     225.91       244.53       1.08x
BenchmarkMultiple/Decode/r30t60/8191KB-8     305.00       279.69       0.92x
```

### CODE REVIEW

File: common.go [hunk 1]
```diff
@@ -35,8 +35,14 @@
...
+	"github.com/spacemonkeygo/errors"
+	"golang.org/x/sys/cpu"
+)
 
 var Error = errors.NewClass("infectious")
 var NotEnoughShares = Error.NewClass("not enough shares")
 var TooManyErrors = Error.NewClass("too many errors")
+
+var hasAVX2 = cpu.X86.HasAVX2
+var hasSSSE3 = cpu.X86.HasSSSE3
```
Comment (zeebo) [inline common.go]: yay!

**Comment by egonelbre** [comment]
There's this part https://github.com/vivint/infectious/blob/master/addmul_test.go#L38

**Comment by egonelbre** [comment]
For whatever reason I didn't get c2goasm work so I hand-rolled, might retry some other day.

**Comment by zeebo** [comment]
Ah indeed. If it stays the same against the trivial Go code, then that's sufficient.

**Pull request merged**